src/librbd/crypto/FlattenRequest.cc

```diff
@@ -0,0 +1,248 @@
+#include <boost/lambda/construct.hpp>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::crypto::FlattenRequest: " << this \
+                           << " " << __func__ << ": "
+
+namespace librbd {
+namespace crypto {
+
+using librbd::util::create_context_callback;
+
+template <typename I>
+class C_FlattenObject : public C_AsyncObjectThrottle<I> {
+public:
+  C_FlattenObject(AsyncObjectThrottle<I> &throttle, I *image_ctx,
+                  IOContext io_context, uint64_t object_no)
+    : C_AsyncObjectThrottle<I>(throttle, *image_ctx), m_io_context(io_context),
+      m_object_no(object_no) {
+  }
+
+  int send() override {
+    if (this->m_image_ctx.object_map != nullptr &&
+        !this->m_image_ctx.object_map->object_may_not_exist(m_object_no)) {
+      // can skip because the object already exists
+      return 1;
+    }
+
+    if (!io::util::trigger_copyup(
+            &this->m_image_ctx, m_object_no, m_io_context, this)) {
+      // stop early if the parent went away - it just means
+      // another flatten finished first or the image was resized
+      return 1;
+    }
+
+    return 0;
+  }
+
+private:
+  IOContext m_io_context;
+  uint64_t m_object_no;
+};
+
+template <typename I>
+FlattenRequest<I>::FlattenRequest(
+        I* image_ctx, Context* on_finish) : m_image_ctx(image_ctx),
+                                            m_raw_image_ctx(nullptr),
+                                            m_on_finish(on_finish) {
+}
+
+template <typename I>
+void FlattenRequest<I>::send() {
+  m_image_ctx->image_lock.lock_shared();
+  bool is_encryption_loaded = m_image_ctx->encryption_format.get() != nullptr;
+  m_image_ctx->image_lock.unlock_shared();
+  if (!is_encryption_loaded) {
+    // no encryption
+    finish(0);
+    return;
+  }
+
+  m_raw_image_ctx = I::create(m_image_ctx->name, m_image_ctx->id, nullptr,
+                              m_image_ctx->data_ctx, false);
+
+  auto ctx = create_context_callback<
+          FlattenRequest<I>, &FlattenRequest<I>::handle_create_raw_ctx>(this);
+  m_raw_image_ctx->state->open(0, ctx);
+}
+
+template <typename I>
+void FlattenRequest<I>::handle_create_raw_ctx(int r) {
+  ldout(m_image_ctx->cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_image_ctx->cct) << "unable to open raw image context: "
+                            << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  ceph_assert(m_image_ctx->exclusive_lock == nullptr ||
+              m_image_ctx->exclusive_lock->is_lock_owner());
+  if (m_raw_image_ctx->exclusive_lock != nullptr) {
+    m_raw_image_ctx->exclusive_lock->unset_require_lock(io::DIRECTION_BOTH);
+  }
+
+  flatten_crypto_header();
+}
+
+template <typename I>
+void FlattenRequest<I>::flatten_crypto_header() {
+  m_image_ctx->image_lock.lock_shared();
+  bool is_encryption_loaded = m_image_ctx->encryption_format.get() != nullptr;
+  uint64_t data_offset;
+  if (is_encryption_loaded) {
```

Comment:
It was a best effort for protecting against a racing shut-down crypto request (perhaps possible via a concurrent format?)

Reply:
Since this doesn't protect from anything like that in full, let's just call any potential code doing that ugly and broken and get rid of these checks (there is another one below).